File: sdk/demo/python/src/prohand_demo/ping.py
#!/usr/bin/env python3
"""
ProHand SDK FFI Demo: Ping Command

Connects to the IPC host and sends periodic ping commands.
"""

import sys
import argparse
import time
from .utils import DemoBase


class PingDemo(DemoBase):
    """Ping command demo."""

    def __init__(self):
        super().__init__("ProHand Ping Demo")

    def run(
        self, command_endpoint: str, status_endpoint: str, hand_streaming_endpoint: str, wrist_streaming_endpoint: str, count: int, interval: float
    ) -> int:
        """Run the ping demo."""
        self.banner()
        print(f"\nCommand endpoint:       {command_endpoint}")
        print(f"Status endpoint:         {status_endpoint}")
        print(f"Hand streaming endpoint: {hand_streaming_endpoint}")
        print(f"Wrist streaming endpoint: {wrist_streaming_endpoint}")
        print(f"Ping count:              {count if count > 0 else 'infinite'}")
        print(f"Interval:                {interval}s")

        try:
            self.section("Connecting...")
            client = self.sdk.ProHandClient(command_endpoint, status_endpoint, hand_streaming_endpoint, wrist_streaming_endpoint)
            self.success("Connected!\n")

            ping_count = 0
            start_time = time.time()

            try:
                while count == 0 or ping_count < count:
                    client.send_ping()
                    ping_count += 1
                    elapsed = time.time() - start_time
                    print(f"[{elapsed:6.2f}s] Ping #{ping_count} sent ✓")

                    if count > 0 and ping_count >= count:
                        break

                    time.sleep(interval)

            except KeyboardInterrupt:
                print("\n")
                self.info("Interrupted by user")
            finally:
                client.close()

            self.success(f"Sent {ping_count} ping(s) successfully")
            return 0

        except self.sdk.ConnectionError as e:
            self.error(f"Connection failed: {e}")
            print("\nMake sure prohand-headless-ipc-host is running")
            return 1
        except self.sdk.ProHandError as e:
            self.error(f"Command failed: {e}")
            return 1
        except Exception as e:
            self.error(f"Unexpected error: {e}")
            import traceback
            traceback.print_exc()
            return 1


def main() -> int:
    """Main entry point."""
    parser = argparse.ArgumentParser(
        description="Send ping commands to ProHand IPC host",
        formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--command-endpoint",
        type=str,
        default="ipc:///tmp/prohand-commands.ipc",
        help="ZMQ command endpoint"
    )
    parser.add_argument(
        "--status-endpoint",
        type=str,
        default="ipc:///tmp/prohand-status.ipc",
        help="ZMQ status endpoint"
    )
    parser.add_argument(
        "--hand-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-hand-streaming.ipc",
        help="ZMQ hand streaming endpoint"
    )
    parser.add_argument(
        "--wrist-streaming-endpoint",
        type=str,
        default="ipc:///tmp/prohand-wrist-streaming.ipc",
        help="ZMQ wrist streaming endpoint"
    )
    parser.add_argument(
        "--count", type=int, default=10, help="Number of pings to send (0 = infinite)"
    )
    parser.add_argument(
        "--interval",
        type=float,
        default=1.0,
        help="Interval between pings (seconds)"
    )
    args = parser.parse_args()

    demo = PingDemo()
    return demo.run(
        args.command_endpoint, args.status_endpoint, args.hand_streaming_endpoint, args.wrist_streaming_endpoint, args.count, args.interval
    )


if __name__ == "__main__":
    sys.exit(main())

File: sdk/demo/cpp/src/ping.cpp
/**
 * ProHand SDK FFI Demo: Ping Command
 */

#include <iostream>
#include <cxxopts.hpp>
#include <chrono>
#include <thread>

#include <prohand_sdk/ProHandClient.hpp>
#include "prohand_demo/Utils.hpp"

using namespace prohand_sdk;
using namespace prohand_demo;  // For utils namespace

int main(int argc, char** argv) {
    cxxopts::Options options("ping", "Send ping commands to ProHand IPC host");

    options.add_options()("command-endpoint", "ZMQ command endpoint",
                          cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-commands.ipc"))("status-endpoint", "ZMQ status endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-status.ipc"))("hand-streaming-endpoint", "ZMQ hand streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-hand-streaming.ipc"))("wrist-streaming-endpoint", "ZMQ wrist streaming endpoint",
                                                                                                           cxxopts::value<std::string>()->default_value("ipc:///tmp/prohand-wrist-streaming.ipc"))("c,count", "Number of pings to send",
                                                                                                                                                                                          cxxopts::value<int>()->default_value("10"))("d,delay", "Delay between pings (seconds)",
                                                                                                                                                                                                                                      cxxopts::value<double>()->default_value("1.0"))("h,help", "Print usage");

    auto result = options.parse(argc, argv);
    
    if (result.count("help")) {
        std::cout << options.help() << std::endl;
        return 0;
    }

    auto commandEndpoint = result["command-endpoint"].as<std::string>();
    auto statusEndpoint = result["status-endpoint"].as<std::string>();
    auto handStreamingEndpoint = result["hand-streaming-endpoint"].as<std::string>();
    auto wristStreamingEndpoint = result["wrist-streaming-endpoint"].as<std::string>();
    auto count = result["count"].as<int>();
    auto delay = result["delay"].as<double>();

    utils::printBanner("ProHand Ping Demo");

    std::cout << "Configuration:\n";
    std::cout << "  Command endpoint:       " << commandEndpoint << "\n";
    std::cout << "  Status endpoint:         " << statusEndpoint << "\n";
    std::cout << "  Hand streaming endpoint: " << handStreamingEndpoint << "\n";
    std::cout << "  Wrist streaming endpoint: " << wristStreamingEndpoint << "\n";
    std::cout << "  Ping count: " << count << "\n";
    std::cout << "  Delay: " << delay << "s\n";

    try {
        utils::printSection("Connecting to IPC host...");
        ProHandClient client(commandEndpoint, statusEndpoint, handStreamingEndpoint, wristStreamingEndpoint);
        utils::printSuccess("Connected!");

        utils::printSection("Sending ping commands...");

        int successCount = 0;
        for (int i = 0; i < count; ++i)
        {
            try
            {
                auto start = std::chrono::steady_clock::now();
                client.sendPing();
                auto end = std::chrono::steady_clock::now();
                auto latency = std::chrono::duration_cast<std::chrono::microseconds>(end - start);

                successCount++;
                std::cout << "Ping " << (i + 1) << "/" << count
                          << " - ✓ Success (latency: " << latency.count() << "µs)\n";

                if (i < count - 1)
                {
                    std::this_thread::sleep_for(
                        std::chrono::duration<double>(delay));
                }
            }
            catch (const SdkException &e)
            {
                std::cout << "Ping " << (i + 1) << "/" << count
                          << " - ✗ Failed: " << e.what() << "\n";
            }
        }

        std::cout << "\n";
        utils::printSuccess("Ping test completed!");
        std::cout << "  Success rate: " << successCount << "/" << count
                  << " (" << (100.0 * successCount / count) << "%)\n";

        return successCount == count ? 0 : 1;
    } catch (const SdkException& e) {
        utils::printError(std::string("Connection failed: ") + e.what());
        std::cout << "\nMake sure the ProHand IPC host is running.\n";
        return 1;
    } catch (const std::exception& e) {
        utils::printError(std::string("Unexpected error: ") + e.what());
        return 1;
    }
}
